timesync/peersync/sync.go

```diff
@@ -0,0 +1,330 @@
+		MaxOffsetErrors:     10,
+		RequiredResponses:   3,
+		ResponsesBufferSize: 20,
+	}
+}
+
+// Config for Sync.
+type Config struct {
+	Disable            bool          `mapstructure:"disable"`
+	RoundRetryInterval time.Duration `mapstructure:"round-retry-interval"`
+	RoundInterval      time.Duration `mapstructure:"round-interval"`
+	RoundTimeout       time.Duration `mapstructure:"round-timeout"`
+	MaxClockOffset     time.Duration `mapstructure:"max-clock-offset"`
+	MaxOffsetErrors    int           `mapstructure:"max-offset-errors"`
+	RequiredResponses  int           `mapstructure:"required-responses"`
+	// ResponsesBufferSize should be updated from p2pconfig.BufferSize
+	ResponsesBufferSize int
+}
+
+// Option to modify Sync behavior.
+type Option func(*Sync)
+
+// WithTime modifies source of time used in Sync.
+func WithTime(t Time) Option {
+	return func(s *Sync) {
+		s.time = t
+	}
+}
+
+// WithContext modifies parent context that is used for all operations in Sync.
+func WithContext(ctx context.Context) Option {
+	return func(s *Sync) {
+		s.ctx = ctx
+	}
+}
+
+// WithLog modifies Log used in Sync.
+func WithLog(lg log.Log) Option {
+	return func(s *Sync) {
+		s.log = lg
+	}
+}
+
+// WithConfig modifies config used in Sync.
+func WithConfig(config Config) Option {
+	return func(s *Sync) {
+		s.config = config
+	}
+}
+
+// New creates Sync instance and returns pointer.
+func New(network Network, opts ...Option) *Sync {
+	sync := &Sync{
+		log:     log.NewNop(),
+		ctx:     context.Background(),
+		time:    systemTime{},
+		config:  DefaultConfig(),
+		network: network,
+		peersWatcher: peersWatcher{
+			requests: make(chan *waitPeersReq),
+		},
+	}
+	for _, opt := range opts {
+		opt(sync)
+	}
+
+	sync.ctx, sync.cancel = context.WithCancel(sync.ctx)
+	sync.wg, sync.ctx = errgroup.WithContext(sync.ctx)
+
+	sync.srv = server.NewMsgServer(sync.ctx,
+		network,
+		protocolName,
+		sync.config.RoundTimeout,
+		make(chan service.DirectMessage, sync.config.ResponsesBufferSize),
+		sync.log,
+	)
+	sync.srv.RegisterBytesMsgHandler(server.RequestTimeSync, sync.requestHandler)
+	return sync
+}
+
+// Sync manages background worker that compares peers time with system time.
+type Sync struct {
+	errCnt uint32
+
+	config       Config
+	log          log.Log
+	srv          *server.MessageServer
+	time         Time
+	network      Network
+	peersWatcher peersWatcher
+
+	once   sync.Once
+	wg     *errgroup.Group
+	ctx    context.Context
+	cancel func()
+}
+
+func (s *Sync) requestHandler(ctx context.Context, buf []byte) []byte {
+	var request Request
+	if err := types.BytesToInterface(buf, &request); err != nil {
+		s.log.Debug("can't decode request", log.Binary("request", buf), log.Err(err))
+		return nil
+	}
+	resp := Response{
+		ID:        request.ID,
+		Timestamp: s.time.Now().UnixNano(),
+	}
+	buf, err := types.InterfaceToBytes(&resp)
+	if err != nil {
+		s.log.Panic("can't encode response", log.Binary("response", buf), log.Err(err))
```

Comment:
should we panic here if response can't be encoded? maybe an error will suffice

Reply:
pretty sure we should panic. this error can be occur only because of the programmer error